client/vaultclient/vaultclient.go

```diff
@@ -0,0 +1,716 @@
+	"github.com/hashicorp/nomad/nomad"
+	"github.com/hashicorp/nomad/nomad/structs"
+	"github.com/hashicorp/nomad/nomad/structs/config"
+	vaultapi "github.com/hashicorp/vault/api"
+)
+
+type TokenDeriverFunc func(*structs.Allocation, []string, *vaultapi.Client) (map[string]string, error)
+
+// The interface which nomad client uses to interact with vault and
+// periodically renews the tokens and secrets.
+type VaultClient interface {
+	// Starts the renewal loop of tokens and secrets
+	Start()
+
+	// Stops the renewal loop for tokens and secrets
+	Stop()
+
+	// Contacts the nomad server and fetches wrapped tokens for a set of
+	// tasks. The wrapped tokens will be unwrapped using vault and
+	// returned.
+	DeriveToken(*structs.Allocation, []string) (map[string]string, error)
+
+	// Fetch the Consul ACL token required for the task
+	GetConsulACL(string, string) (*vaultapi.Secret, error)
+
+	// Renews a token with the given increment and adds it to the min-heap
+	// for periodic renewal.
+	RenewToken(string, int) <-chan error
+
+	// Removes the token from the min-heap, stopping its renewal.
+	StopRenewToken(string) error
+
+	// Renews a vault secret's lease and add the lease identifier to the
+	// min-heap for periodic renewal.
+	RenewLease(string, int) <-chan error
+
+	// Removes a secret's lease ID from the min-heap, stopping its renewal.
+	StopRenewLease(string) error
+}
+
+// Implementation of VaultClient interface to interact with vault and perform
+// token and lease renewals periodically.
+type vaultClient struct {
+	tokenDeriver TokenDeriverFunc
+
+	// running indicates if the renewal loop is active or not
+	running bool
+
+	// connEstablished marks whether the connection to vault was successful
+	// or not
+	connEstablished bool
+
+	// tokenData is the data of the passed VaultClient token
+	token *tokenData
+
+	// API client to interact with vault
+	client *vaultapi.Client
+
+	// Channel to notify heap modifications to the renewal loop
+	updateCh chan struct{}
+
+	// Channel to trigger termination of renewal loop
+	stopCh chan struct{}
+
+	// Min-Heap to keep track of both tokens and leases
+	heap *vaultClientHeap
+
+	// Configuration to connect to vault
+	config *config.VaultConfig
+
+	lock   sync.RWMutex
+	logger *log.Logger
+
+	rpcHandler clientconfig.RPCHandler
+	rpcProxy   *rpcproxy.RPCProxy
+	connPool   *nomad.ConnPool
+}
+
+// tokenData holds the relevant information about the Vault token passed to the
+// client.
+type tokenData struct {
+	CreationTTL int      `mapstructure:"creation_ttl"`
+	TTL         int      `mapstructure:"ttl"`
+	Renewable   bool     `mapstructure:"renewable"`
+	Policies    []string `mapstructure:"policies"`
+	Role        string   `mapstructure:"role"`
+	Root        bool
+}
+
+// Request object for renewals. This can be used for both token renewals and
+// secret's lease renewals.
+type vaultClientRenewalRequest struct {
+	// Channel into which any renewal error will be sent down to
```

Comment:
You don't have to update these but typical style for comments is to start with the name of the type/field/function:
https://golang.org/doc/effective_go.html#commentary